### ISSUE TITLE
Warnings not thrown when convert some JSON value to double

### ISSUE THREAD

**Issue by winoros** [pingcap/tidb#10554]
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?
If possible, provide a recipe for reproducing the error.

```
CREATE TABLE tj2(j JSON, i INT DEFAULT 7);
INSERT INTO tj2(j) VALUES ('1');
INSERT INTO tj2(j) VALUES ('2');
INSERT INTO tj2(j) VALUES ('3');
INSERT INTO tj2(j) VALUES ('4');
INSERT INTO tj2(j) VALUES ('5');
INSERT INTO tj2(j) VALUES (NULL);
INSERT INTO tj2(j) VALUES ('3.14');
INSERT INTO tj2(j) VALUES ('[1,2,3]');
SELECT CAST(SUM(j) OVER () AS JSON) FROM tj;
```

2. What did you expect to see?

executes successfully and throw a waring `Invalid JSON value for CAST to DOUBLE from column j at row 8`.

3. What did you see instead?

yes it returns correct data set, but with no warning.

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?

**Comment by winoros** [comment]
```
SELECT j,CAST(SUM(j) OVER (PARTITION BY i) AS JSON), CAST(SUM(j) OVER () AS JSON) FROM tj;
Warning 3156    Invalid JSON value for CAST to DOUBLE from column j at row 3
Warning 3156    Invalid JSON value for CAST to DOUBLE from column j at row 3
```

**Comment by winoros** [comment]
Also the flen is not correct.
```
UPDATE tj SET i=i+CASE WHEN JSON_TYPE(j) = 'ARRAY' THEN 1 ELSE j END;
UPDATE tj SET i=7 where i=8 AND JSON_TYPE(j) != 'ARRAY';
SELECT j,CAST(SUM(j) OVER (PARTITION BY i) AS JSON), CAST(SUM(j) OVER () AS JSON) FROM tj;
```

Result of TiDB:
```
+-----------+--------------------------------------------+------------------------------+
| j         | CAST(SUM(j) OVER (PARTITION BY i) AS JSON) | CAST(SUM(j) OVER () AS JSON) |
+-----------+--------------------------------------------+------------------------------+
| NULL      | NULL                                       | 18.14                        |
| 1         | 1                                          | 18.14                        |
| [1, 2, 3] | 0                                          | 18.14                        |
| 2         | 2                                          | 18.14                        |
| 3         | 6.140000000000001                          | 18.14                        |
| 3.14      | 6.140000000000001                          | 18.14                        |
| 4         | 4                                          | 18.14                        |
| 5         | 5                                          | 18.14                        |
+-----------+--------------------------------------------+------------------------------+
```

Result of MySQL:
```
+-----------+--------------------------------------------+------------------------------+
| j         | CAST(SUM(j) OVER (PARTITION BY i) AS JSON) | CAST(SUM(j) OVER () AS JSON) |
+-----------+--------------------------------------------+------------------------------+
| NULL      | NULL                                       | 18.14                        |
| 1         | 1.0                                        | 18.14                        |
| [1, 2, 3] | 0.0                                        | 18.14                        |
| 2         | 2.0                                        | 18.14                        |
| 3         | 6.140000000000001                          | 18.14                        |
| 3.14      | 6.140000000000001                          | 18.14                        |
| 4         | 4.0                                        | 18.14                        |
| 5         | 5.0                                        | 18.14                        |
+-----------+--------------------------------------------+------------------------------+
```

**Comment by nullnotnil** [comment]
It looks like the warnings are still thrown in MySQL, but not TiDB:

```sql
DROP TABLE IF EXISTS tj;
CREATE TABLE tj(j JSON, i INT DEFAULT 7);
INSERT INTO tj(j) VALUES ('1');
INSERT INTO tj(j) VALUES ('2');
INSERT INTO tj(j) VALUES ('3');
INSERT INTO tj(j) VALUES ('4');
INSERT INTO tj(j) VALUES ('5');
INSERT INTO tj(j) VALUES (NULL);
INSERT INTO tj(j) VALUES ('3.14');
INSERT INTO tj(j) VALUES ('[1,2,3]');
UPDATE tj SET i=i+CASE WHEN JSON_TYPE(j) = 'ARRAY' THEN 1 ELSE j END;
UPDATE tj SET i=7 where i=8 AND JSON_TYPE(j) != 'ARRAY';
SELECT j,CAST(SUM(j) OVER (PARTITION BY i) AS JSON), CAST(SUM(j) OVER () AS JSON) FROM tj ORDER BY j;

..

mysql> SELECT j,CAST(SUM(j) OVER (PARTITION BY i) AS JSON), CAST(SUM(j) OVER () AS JSON) FROM tj ORDER BY j;
+-----------+--------------------------------------------+------------------------------+
| j         | CAST(SUM(j) OVER (PARTITION BY i) AS JSON) | CAST(SUM(j) OVER () AS JSON) |
+-----------+--------------------------------------------+------------------------------+
| NULL      | NULL                                       | 18.14                        |
| 1         | 1                                          | 18.14                        |
| 2         | 2                                          | 18.14                        |
| 3         | 6.140000000000001                          | 18.14                        |
| 3.14      | 6.140000000000001                          | 18.14                        |
| 4         | 4                                          | 18.14                        |
| 5         | 5                                          | 18.14                        |
| [1, 2, 3] | 0                                          | 18.14                        |
+-----------+--------------------------------------------+------------------------------+
8 rows in set (0.00 sec)

mysql> SELECT tidb_version()\G
*************************** 1. row ***************************
tidb_version(): Release Version: v4.0.0-beta.2-870-g2a8b96845
Edition: Community
Git Commit Hash: 2a8b968453520e4fcf9d6ff46c9f23b4ad23feee
Git Branch: master
UTC Build Time: 2020-07-31 08:45:35
GoVersion: go1.13
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
1 row in set (0.00 sec)
```


Vs MySQL 8.0.21:
```sql
mysql [localhost:8021] {msandbox} (test) > SELECT j,CAST(SUM(j) OVER (PARTITION BY i) AS JSON), CAST(SUM(j) OVER () AS JSON) FROM tj ORDER BY j;
+-----------+--------------------------------------------+------------------------------+
| j         | CAST(SUM(j) OVER (PARTITION BY i) AS JSON) | CAST(SUM(j) OVER () AS JSON) |
+-----------+--------------------------------------------+------------------------------+
| NULL      | NULL                                       | 18.14                        |
| 1         | 1.0                                        | 18.14                        |
| 2         | 2.0                                        | 18.14                        |
| 3         | 6.140000000000001                          | 18.14                        |
| 3.14      | 6.140000000000001                          | 18.14                        |
| 4         | 4.0                                        | 18.14                        |
| 5         | 5.0                                        | 18.14                        |
| [1, 2, 3] | 0.0                                        | 18.14                        |
+-----------+--------------------------------------------+------------------------------+
8 rows in set, 3 warnings (0.00 sec)

mysql [localhost:8021] {msandbox} (test) > SHOW WARNINGS\G
*************************** 1. row ***************************
  Level: Warning
   Code: 3156
Message: Invalid JSON value for CAST to DOUBLE from column j at row 1
*************************** 2. row ***************************
  Level: Warning
   Code: 3156
Message: Invalid JSON value for CAST to DOUBLE from column j at row 1
*************************** 3. row ***************************
  Level: Warning
   Code: 1235
Message: This version of MySQL doesn't yet support 'sorting of non-scalar JSON values'
3 rows in set (0.00 sec)

```

**Comment by YangKeao** [comment]
```
mysql> CREATE TABLE tj2(j JSON, i INT DEFAULT 7);
Query OK, 0 rows affected (0.11 sec)

mysql> INSERT INTO tj2(j) VALUES ('1');
Query OK, 1 row affected (0.00 sec)

mysql> INSERT INTO tj2(j) VALUES ('2');
Query OK, 1 row affected (0.01 sec)

mysql> INSERT INTO tj2(j) VALUES ('3');
Query OK, 1 row affected (0.00 sec)

mysql> INSERT INTO tj2(j) VALUES ('4');
Query OK, 1 row affected (0.00 sec)

mysql> INSERT INTO tj2(j) VALUES ('5');
Query OK, 1 row affected (0.00 sec)

mysql> INSERT INTO tj2(j) VALUES (NULL);
Query OK, 1 row affected (0.00 sec)

mysql> INSERT INTO tj2(j) VALUES ('3.14');
Query OK, 1 row affected (0.01 sec)

mysql> INSERT INTO tj2(j) VALUES ('[1,2,3]');
Query OK, 1 row affected (0.00 sec)

mysql> SELECT CAST(SUM(j) OVER () AS JSON) FROM tj;
ERROR 1146 (42S02): Table 'test.tj' doesn't exist
mysql> SELECT CAST(SUM(j) OVER () AS JSON) FROM tj2;
+------------------------------+
| CAST(SUM(j) OVER () AS JSON) |
+------------------------------+
| 18.14                        |
| 18.14                        |
| 18.14                        |
| 18.14                        |
| 18.14                        |
| 18.14                        |
| 18.14                        |
| 18.14                        |
+------------------------------+
8 rows in set, 1 warning (0.00 sec)

mysql> show warnings;
+---------+------+----------------------------------------------+
| Level   | Code | Message                                      |
+---------+------+----------------------------------------------+
| Warning | 1292 | Truncated incorrect FLOAT value: '[1, 2, 3]' |
+---------+------+----------------------------------------------+
1 row in set (0.00 sec)
```

It seems that this issue has been fixed at least in master and 6.5.
